src/stacked_borrows.rs

```diff
@@ -71,15 +71,35 @@ impl fmt::Debug for Item {
 }
 
 /// Extra per-location state.
-#[derive(Clone, Debug, PartialEq, Eq)]
+#[derive(Clone, Debug)]
 pub struct Stack {
     /// Used *mostly* as a stack; never empty.
     /// Invariants:
     /// * Above a `SharedReadOnly` there can only be more `SharedReadOnly`.
     /// * Except for `Untagged`, no tag occurs in the stack more than once.
     borrows: Vec<Item>,
+    /// We often want to know what (if any) tag in the stack exists for some `SbTag`.
+    /// The tag we're looking for does not occur often enough at the end of the stack for a linear
+    /// search to be efficient, so we use the uniqueness guarantee to keep a map from tag to
+    /// position in the stack.
+    cache: FxHashMap<PtrId, usize>,
+    /// `Untagged` may occur multiple times so we store it outside of the map.
+    untagged: Vec<usize>,
```

Comment:
Perhaps this is pre-existing, but it might be a good idea to use a newtype here instead of `usize`.

Reply:
How would that help? These values exist only to be used as indexes into `borrows`.

Reply:
Ah, then it'd be good to add a comment about that. Reading this field definition, I had a couple of confusion points:

- What is the meaning of the `usize`?
- Is this ordered or could it be a set instead?

Reply:
Your newtype suggestion is looking better and better as I try things out :sweat_smile: